File: infra/docker/docker.go
package docker

import (
	"fmt"
	"lunch/domain/port"
	"os"
	"os/exec"
	"path/filepath"
)

// Builder implements the DockerBuilder interface.
type Builder struct{}

// NewBuilder creates a new Docker builder.
func NewBuilder() port.DockerBuilder {
	return &Builder{}
}

// BuildImage builds a Docker image from the specified Dockerfile.
// dockerFile is the path to the Dockerfile.
// outputDir is the build context directory where the image will be built from.
func (b *Builder) BuildImage(dockerFile string, outputDir string) error {
	// Verify the Dockerfile exists
	if _, err := os.Stat(dockerFile); os.IsNotExist(err) {
		return fmt.Errorf("dockerfile not found: %s", dockerFile)
	}

	// Verify the output directory exists
	if _, err := os.Stat(outputDir); os.IsNotExist(err) {
		return fmt.Errorf("output directory not found: %s", outputDir)
	}

	// Get the image name from the directory name
	imageName := filepath.Base(outputDir)

	// Build the Docker image
	cmd := exec.Command("docker", "build", "-f", dockerFile, "-t", imageName, outputDir)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to build docker image: %w", err)
	}

	return nil
}

File: src/Infra/Docker.hs
{-# LANGUAGE OverloadedStrings #-}

module Infra.Docker (newDockerRepo) where

import Control.Monad (unless)
import Control.Monad.Trans.Class (lift)
import Control.Monad.Trans.Except (throwE)
import Data.Aeson (Value, decode, withObject, (.:))
import Data.Aeson.Types (Parser, parseMaybe)
import qualified Data.ByteString.Lazy.Char8 as BSL
import Data.Maybe (listToMaybe)
import Domain.Model (AppError (..), AppInfo (AppInfo), AppStatus (..), ContainerInfo (..), Result)
import Domain.Port (DockerRepo (..), Logger (..))
import Pkg.IO (tryCmd)
import System.Directory (doesFileExist)
import System.FilePath (takeBaseName, (</>))

-- | Assert that both Dockerfile and docker-compose.yml exist in @path@.
checkDockerFiles :: FilePath -> Result ()
checkDockerFiles path = do
  fileMustExist (path </> "Dockerfile") "Missing Dockerfile"
  fileMustExist (path </> "docker-compose.yml") "Missing docker-compose.yml"

-- | Asserts that @file@ exists, returning @error message@ if not.
fileMustExist :: FilePath -> String -> Result ()
fileMustExist path msg = do
  exists <- lift $ doesFileExist path
  unless exists $ throwE (ConfigError msg)

-- | Run compose using @logger@ to print @info@.
-- Runs command with @args@ and returns @error@ on failure.
runCompose :: Logger -> String -> [String] -> AppError -> FilePath -> Result ()
runCompose logger msg args err dir = do
  checkDockerFiles dir
  lift $ logInfo logger (msg <> takeBaseName dir <> "...")
  result <- tryCmd "docker" (["compose", "--project-directory", dir] <> args)
  case result of
    Left e -> do
      lift $ logError logger (show e)
      throwE err
    Right _ -> pure ()

-- | Parse json output from Docker inspect command into ContainerInfo.
parseInspectJson :: String -> Maybe ContainerInfo
parseInspectJson raw = do
  arr <- decode (BSL.pack raw) :: Maybe [Value]
  val <- listToMaybe arr
  parseMaybe parseContainer val

parseContainer :: Value -> Parser ContainerInfo
parseContainer = withObject "container" $ \o -> do
  name <- o .: "Name"
  state <- o .: "State"
  status <- state .: "Status"
  running <- state .: "Running"
  config <- o .: "Config"
  image <- config .: "Image"
  pure
    ContainerInfo
      { cName = name,
        cStatus = status,
        cRunning = running,
        cImage = image
      }

newDockerRepo :: Logger -> DockerRepo
newDockerRepo logger =
  DockerRepo
    { buildProject = runCompose logger "Building Docker image for " ["up", "--build", "-d"] (DockerError "Docker build failed"),
      composeDown = runCompose logger "Stopping " ["down"] (DockerError "Docker compose down failed"),
      getCid = \fp -> do
        result <- tryCmd "docker" ["compose", "--project-directory", fp, "ps", "-q"]
        case result of
          Left e -> do
            lift $ logError logger (show e)
            throwE (DockerError "Failed to read container id")
          Right cid -> pure $ if cid == "" then Nothing else Just cid,
      getInfo = \name cid -> do
        result <- tryCmd "docker" ["inspect", cid]
        maybeInfo <- case result of
          Left e -> do
            lift $ logError logger (show e)
            throwE (DockerError "Failed to inspect container")
          Right raw -> pure $ parseInspectJson raw
        case maybeInfo of
          Just cinfo -> pure (AppInfo name (if cRunning cinfo then Running else Stopped))
          Nothing -> throwE $ DockerError "Failed to parse inspect output"
    }
